File: app/src/androidTest/java/com/todoroo/astrid/data/TaskTest.java
package com.todoroo.astrid.data;

import static com.todoroo.astrid.data.Task.URGENCY_DAY_AFTER;
import static com.todoroo.astrid.data.Task.URGENCY_IN_TWO_WEEKS;
import static com.todoroo.astrid.data.Task.URGENCY_NEXT_MONTH;
import static com.todoroo.astrid.data.Task.URGENCY_NEXT_WEEK;
import static com.todoroo.astrid.data.Task.URGENCY_NONE;
import static com.todoroo.astrid.data.Task.URGENCY_SPECIFIC_DAY;
import static com.todoroo.astrid.data.Task.URGENCY_SPECIFIC_DAY_TIME;
import static com.todoroo.astrid.data.Task.URGENCY_TODAY;
import static com.todoroo.astrid.data.Task.URGENCY_TOMORROW;
import static com.todoroo.astrid.data.Task.createDueDate;
import static com.todoroo.astrid.data.Task.hasDueTime;
import static junit.framework.Assert.assertEquals;
import static junit.framework.Assert.assertFalse;
import static junit.framework.Assert.assertTrue;
import static org.tasks.Freeze.freezeAt;
import static org.tasks.Freeze.thaw;
import static org.tasks.date.DateTimeUtils.newDateTime;
import static org.tasks.time.DateTimeUtils.currentTimeMillis;

import android.support.test.runner.AndroidJUnit4;
import com.todoroo.astrid.data.Task.Priority;
import java.util.ArrayList;
import java.util.TreeSet;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.tasks.Snippet;
import org.tasks.time.DateTime;

@RunWith(AndroidJUnit4.class)
public class TaskTest {

  private static final DateTime now = new DateTime(2013, 12, 31, 16, 10, 53, 452);
  private static final DateTime specificDueDate = new DateTime(2014, 3, 17, 9, 54, 27, 959);

  @Before
  public void setUp() {
    freezeAt(now);
  }

  @After
  public void tearDown() {
    thaw();
  }

  @Test
  public void testCreateDueDateNoUrgency() {
    assertEquals(0, createDueDate(URGENCY_NONE, 1L));
  }

  @Test
  public void testCreateDueDateToday() {
    long expected = new DateTime(2013, 12, 31, 12, 0, 0, 0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_TODAY, -1L));
  }

  @Test
  public void testCreateDueDateTomorrow() {
    long expected = new DateTime(2014, 1, 1, 12, 0, 0, 0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_TOMORROW, -1L));
  }

  @Test
  public void testCreateDueDateDayAfter() {
    long expected = new DateTime(2014, 1, 2, 12, 0, 0, 0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_DAY_AFTER, -1L));
  }

  @Test
  public void testCreateDueDateNextWeek() {
    long expected = new DateTime(2014, 1, 7, 12, 0, 0, 0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_NEXT_WEEK, -1L));
  }

  @Test
  public void testCreateDueDateInTwoWeeks() {
    long expected = new DateTime(2014, 1, 14, 12, 0, 0, 0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_IN_TWO_WEEKS, -1L));
  }

  @Test
  public void testCreateDueDateNextMonth() {
    long expected = new DateTime(2014, 1, 31, 12, 0, 0, 0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_NEXT_MONTH, -1L));
  }

  @Test
  public void testRemoveTimeForSpecificDay() {
    long expected =
        specificDueDate
            .withHourOfDay(12)
            .withMinuteOfHour(0)
            .withSecondOfMinute(0)
            .withMillisOfSecond(0)
            .getMillis();
    assertEquals(expected, createDueDate(URGENCY_SPECIFIC_DAY, specificDueDate.getMillis()));
  }

  @Test
  public void testRemoveSecondsForSpecificTime() {
    long expected = specificDueDate.withSecondOfMinute(1).withMillisOfSecond(0).getMillis();
    assertEquals(expected, createDueDate(URGENCY_SPECIFIC_DAY_TIME, specificDueDate.getMillis()));
  }

  @Test
  public void testTaskHasDueTime() {
    Task task = new Task();
    task.setDueDate(1388516076000L);
    assertTrue(task.hasDueTime());
    assertTrue(task.hasDueDate());
  }

  @Test
  public void testTaskHasDueDate() {
    Task task = new Task();
    task.setDueDate(1388469600000L);
    assertFalse(task.hasDueTime());
    assertTrue(task.hasDueDate());
  }

  @Test
  public void testDoesHaveDueTime() {
    assertTrue(hasDueTime(1388516076000L));
  }

  @Test
  public void testNoDueTime() {
    assertFalse(hasDueTime(newDateTime().startOfDay().getMillis()));
    assertFalse(hasDueTime(newDateTime().withMillisOfDay(60000).getMillis()));
  }

  @Test
  public void testHasDueTime() {
    assertTrue(hasDueTime(newDateTime().withMillisOfDay(1).getMillis()));
    assertTrue(hasDueTime(newDateTime().withMillisOfDay(1000).getMillis()));
    assertTrue(hasDueTime(newDateTime().withMillisOfDay(59999).getMillis()));
  }

  @Test
  public void testDoesNotHaveDueTime() {
    assertFalse(hasDueTime(1388469600000L));
  }

  @Test
  public void testNewTaskIsNotCompleted() {
    assertFalse(new Task().isCompleted());
  }

  @Test
  public void testNewTaskNotDeleted() {
    assertFalse(new Task().isDeleted());
  }

  @Test
  public void testNewTaskNotHidden() {
    assertFalse(new Task().isHidden());
  }

  @Test
  public void testNewTaskDoesNotHaveDueDateOrTime() {
    assertFalse(new Task().hasDueDate());
    assertFalse(new Task().hasDueTime());
  }

  @Test
  public void testTaskIsCompleted() {
    Task task = new Task();
    task.setCompletionDate(1L);
    assertTrue(task.isCompleted());
  }

  @Test
  public void testTaskIsNotHiddenAtHideUntilTime() {
    final long now = currentTimeMillis();
    freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setHideUntil(now);
                assertFalse(task.isHidden());
              }
            });
  }

  @Test
  public void testTaskIsHiddenBeforeHideUntilTime() {
    final long now = currentTimeMillis();
    freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setHideUntil(now + 1);
                assertTrue(task.isHidden());
              }
            });
  }

  @Test
  public void testTaskIsDeleted() {
    Task task = new Task();
    task.setDeletionDate(1L);
    assertTrue(task.isDeleted());
  }

  @Test
  public void testTaskWithNoDueDateIsOverdue() {
    assertTrue(new Task().isOverdue());
  }

  @Test
  public void testTaskNotOverdueAtDueTime() {
    final long now = currentTimeMillis();
    freezeAt(now)
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setDueDate(now);
                assertFalse(task.isOverdue());
              }
            });
  }

  @Test
  public void testTaskIsOverduePastDueTime() {
    final long dueDate = currentTimeMillis();
    freezeAt(dueDate + 1)
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setDueDate(dueDate);
                assertTrue(task.isOverdue());
              }
            });
  }

  @Test
  public void testTaskNotOverdueBeforeNoonOnDueDate() {
    final DateTime dueDate = new DateTime().startOfDay();
    freezeAt(dueDate.plusHours(12).minusMillis(1))
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setDueDate(dueDate.getMillis());
                assertFalse(task.hasDueTime());
                assertFalse(task.isOverdue());
              }
            });
  }

  @Test
  public void testTaskOverdueAtNoonOnDueDate() {
    final DateTime dueDate = new DateTime().startOfDay();
    freezeAt(dueDate.plusHours(12))
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setDueDate(dueDate.getMillis());
                assertFalse(task.hasDueTime());
                assertFalse(task.isOverdue());
              }
            });
  }

  @Test
  public void testTaskWithNoDueTimeIsOverdue() {
    final DateTime dueDate = new DateTime().startOfDay();
    freezeAt(dueDate.plusDays(1))
        .thawAfter(
            new Snippet() {
              {
                Task task = new Task();
                task.setDueDate(dueDate.getMillis());
                assertFalse(task.hasDueTime());
                assertTrue(task.isOverdue());
              }
            });
  }

  @Test
  @SuppressWarnings("ConstantConditions")
  public void testSanity() {
    assertTrue(Priority.HIGH < Priority.MEDIUM);
    assertTrue(Priority.MEDIUM < Priority.LOW);
    assertTrue(Priority.LOW < Priority.NONE);

    ArrayList<Integer> reminderFlags = new ArrayList<>();
    reminderFlags.add(Task.NOTIFY_AFTER_DEADLINE);
    reminderFlags.add(Task.NOTIFY_AT_DEADLINE);
    reminderFlags.add(Task.NOTIFY_MODE_NONSTOP);

    // assert no duplicates
    assertEquals(new TreeSet<>(reminderFlags).size(), reminderFlags.size());
  }
}

File: app/src/test/java/com/todoroo/astrid/data/TaskTest.kt
package com.todoroo.astrid.data

import org.junit.After
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Before
import org.junit.Test
import org.tasks.Freeze
import org.tasks.Freeze.Companion.freezeAt
import org.tasks.data.createDueDate
import org.tasks.data.entity.Task
import org.tasks.data.isHidden
import org.tasks.data.isOverdue
import org.tasks.date.DateTimeUtils
import org.tasks.time.DateTime
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import java.util.TreeSet

class TaskTest {
    @Before
    fun setUp() {
        freezeAt(now)
    }

    @After
    fun tearDown() {
        Freeze.thaw()
    }

    @Test
    fun testCreateDueDateNoUrgency() {
        assertEquals(0, createDueDate(Task.URGENCY_NONE, 1L))
    }

    @Test
    fun testCreateDueDateToday() {
        val expected = DateTime(2013, 12, 31, 12, 0, 0, 0).millis
        assertEquals(expected, createDueDate(Task.URGENCY_TODAY, -1L))
    }

    @Test
    fun testCreateDueDateTomorrow() {
        val expected = DateTime(2014, 1, 1, 12, 0, 0, 0).millis
        assertEquals(expected, createDueDate(Task.URGENCY_TOMORROW, -1L))
    }

    @Test
    fun testCreateDueDateDayAfter() {
        val expected = DateTime(2014, 1, 2, 12, 0, 0, 0).millis
        assertEquals(expected, createDueDate(Task.URGENCY_DAY_AFTER, -1L))
    }

    @Test
    fun testCreateDueDateNextWeek() {
        val expected = DateTime(2014, 1, 7, 12, 0, 0, 0).millis
        assertEquals(expected, createDueDate(Task.URGENCY_NEXT_WEEK, -1L))
    }

    @Test
    fun testCreateDueDateInTwoWeeks() {
        val expected = DateTime(2014, 1, 14, 12, 0, 0, 0).millis
        assertEquals(expected, createDueDate(Task.URGENCY_IN_TWO_WEEKS, -1L))
    }

    @Test
    fun testRemoveTimeForSpecificDay() {
        val expected = specificDueDate
                .withHourOfDay(12)
                .withMinuteOfHour(0)
                .withSecondOfMinute(0)
                .withMillisOfSecond(0)
                .millis
        assertEquals(expected, createDueDate(Task.URGENCY_SPECIFIC_DAY, specificDueDate.millis))
    }

    @Test
    fun testRemoveSecondsForSpecificTime() {
        val expected = specificDueDate.withSecondOfMinute(1).withMillisOfSecond(0).millis
        assertEquals(expected, createDueDate(Task.URGENCY_SPECIFIC_DAY_TIME, specificDueDate.millis))
    }

    @Test
    fun testTaskHasDueTime() {
        val task = Task()
        task.dueDate = 1388516076000L
        assertTrue(task.hasDueTime())
        assertTrue(task.hasDueDate())
    }

    @Test
    fun testTaskHasDueDate() {
        val task = Task()
        task.dueDate = 1388469600000L
        assertFalse(task.hasDueTime())
        assertTrue(task.hasDueDate())
    }

    @Test
    fun testDoesHaveDueTime() {
        assertTrue(Task.hasDueTime(1388516076000L))
    }

    @Test
    fun testNoDueTime() {
        assertFalse(Task.hasDueTime(DateTimeUtils.newDateTime().startOfDay().millis))
        assertFalse(Task.hasDueTime(DateTimeUtils.newDateTime().withMillisOfDay(60000).millis))
    }

    @Test
    fun testHasDueTime() {
        assertTrue(Task.hasDueTime(DateTimeUtils.newDateTime().withMillisOfDay(1).millis))
        assertTrue(Task.hasDueTime(DateTimeUtils.newDateTime().withMillisOfDay(1000).millis))
        assertTrue(Task.hasDueTime(DateTimeUtils.newDateTime().withMillisOfDay(59999).millis))
    }

    @Test
    fun testDoesNotHaveDueTime() {
        assertFalse(Task.hasDueTime(1388469600000L))
    }

    @Test
    fun testNewTaskIsNotCompleted() {
        assertFalse(Task().isCompleted)
    }

    @Test
    fun testNewTaskNotDeleted() {
        assertFalse(Task().isDeleted)
    }

    @Test
    fun testNewTaskNotHidden() {
        assertFalse(Task().isHidden)
    }

    @Test
    fun testNewTaskDoesNotHaveDueDateOrTime() {
        assertFalse(Task().hasDueDate())
        assertFalse(Task().hasDueTime())
    }

    @Test
    fun testTaskIsCompleted() {
        val task = Task()
        task.completionDate = 1L
        assertTrue(task.isCompleted)
    }

    @Test
    fun testTaskIsNotHiddenAtHideUntilTime() {
        val now = currentTimeMillis()
        freezeAt(now) {
            val task = Task()
            task.hideUntil = now
            assertFalse(task.isHidden)
        }
    }

    @Test
    fun testTaskIsHiddenBeforeHideUntilTime() {
        val now = currentTimeMillis()
        freezeAt(now) {
            val task = Task()
            task.hideUntil = now + 1
            assertTrue(task.isHidden)
        }
    }

    @Test
    fun testTaskIsDeleted() {
        val task = Task()
        task.deletionDate = 1L
        assertTrue(task.isDeleted)
    }

    @Test
    fun testTaskWithNoDueDateIsNotOverdue() {
        assertFalse(Task().isOverdue)
    }

    @Test
    fun testTaskNotOverdueAtDueTime() {
        val now = currentTimeMillis()
        freezeAt(now) {
            val task = Task()
            task.dueDate = now
            assertFalse(task.isOverdue)
        }
    }

    @Test
    fun testTaskIsOverduePastDueTime() {
        val dueDate = currentTimeMillis()
        freezeAt(dueDate + 1) {
            val task = Task()
            task.dueDate = dueDate
            assertTrue(task.isOverdue)
        }
    }

    @Test
    fun testTaskNotOverdueBeforeNoonOnDueDate() {
        val dueDate = DateTime().startOfDay()
        freezeAt(dueDate.plusHours(12).minusMillis(1)) {
            val task = Task()
            task.dueDate = dueDate.millis
            assertFalse(task.hasDueTime())
            assertFalse(task.isOverdue)
        }
    }

    @Test
    fun testTaskOverdueAtNoonOnDueDate() {
        val dueDate = DateTime().startOfDay()
        freezeAt(dueDate.plusHours(12)) {
            val task = Task()
            task.dueDate = dueDate.millis
            assertFalse(task.hasDueTime())
            assertFalse(task.isOverdue)
        }
    }

    @Test
    fun testTaskWithNoDueTimeIsOverdue() {
        val dueDate = DateTime().startOfDay()
        freezeAt(dueDate.plusDays(1)) {
            val task = Task()
            task.dueDate = dueDate.millis
            assertFalse(task.hasDueTime())
            assertTrue(task.isOverdue)
        }
    }

    @Test
    fun testSanity() {
        assertTrue(Task.Priority.HIGH < Task.Priority.MEDIUM)
        assertTrue(Task.Priority.MEDIUM < Task.Priority.LOW)
        assertTrue(Task.Priority.LOW < Task.Priority.NONE)
        val reminderFlags = ArrayList<Int>()
        reminderFlags.add(Task.NOTIFY_AFTER_DEADLINE)
        reminderFlags.add(Task.NOTIFY_AT_DEADLINE)
        reminderFlags.add(Task.NOTIFY_MODE_NONSTOP)

        // assert no duplicates
        assertEquals(TreeSet(reminderFlags).size, reminderFlags.size)
    }

    companion object {
        private val now = DateTime(2013, 12, 31, 16, 10, 53, 452)
        private val specificDueDate = DateTime(2014, 3, 17, 9, 54, 27, 959)
    }
}